### PR TITLE
AMP-29169: Reimplement dependency validator

### PR DESCRIPTION
**Pull request by developster**
- Reimplemented dependency validators
- Fixed agreement import
- Upgraded maven-surefire-plugin
- Fixed object importer to validate fields in all cases

### CODE REVIEW

File: amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/SimpleFieldAccessor.java [hunk 1]
```diff
@@ -10,6 +10,9 @@
 /**
  * TODO rename to Reflective
  *
+ * FIXME optimize! searching java.lang.reflect.Field involves involves throwing exceptions which are exceptionally slow
```
Comment (developster) [inline amp/WEB-INF/src/org/digijava/kernel/ampapi/endpoints/activity/SimpleFieldAccessor.java]: Reported AMP-29208.

File: amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        AmpCategoryValue regionalImplementationLevel = categoryValues.getImplementationLevels().getRegional();
+        AmpCategoryValue regionImplementationLocation = categoryValues.getImplementationLocations().getRegion();
+
+        AmpActivity activity = new AmpActivity();
+        activity.setLocations(ImmutableSet.of(haitiUsage));
+        activity.setCategories(ImmutableSet.of(regionalImplementationLevel, regionImplementationLocation));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, contains(locViolation(96L)));
```
Comment (nmandrescu) [inline amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java]: I understand this test for locations, though may be good to also check implLoc violation is reported together.

File: amp/src/test/java/org/digijava/kernel/validators/activity/OnBudgetValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        AmpActivity activity = new AmpActivity();
+        activity.setCategories(ImmutableSet.of(categoryValues.getActivityBudgets().getOnBudget()));
+
+        Set<ConstraintViolation> violations = getConstraintViolations(activity);
+
+        assertThat(violations, emptyIterable());
+    }
+
+    @Test
+    public void testOnBudgetAllMissingFYNotRequired() {
```
Comment (nmandrescu) [inline amp/src/test/java/org/digijava/kernel/validators/activity/OnBudgetValidatorTest.java]: This is something I spent time when checking the ConditionalRequiredValidator that is not causing issues :)))

File: amp/WEB-INF/src/org/digijava/kernel/validators/activity/ConditionalRequiredValidator.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+     * @param type type of the value
+     * @param value value being validated
+     * @return whenever the validator is active or not
+     */
+    public abstract boolean isActive(APIField type, Object value);
+
+    private boolean isEmptyValue(Object value) {
+        return value == null
+                || (value instanceof Collection && ((Collection) value).isEmpty())
+                || (value instanceof String && ((String) value).isEmpty());
```
[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/validators/activity/ConditionalRequiredValidator.java]: Support for translations in validation is not yet implemented. This validator applies currently only to untranslated fields. Added a note that this validator does not support translated fields.

Let's implement support for translations when required validator will be migrated to new validation.

[the remark this answers is not in the document]
Comment (developster) [reply amp/WEB-INF/src/org/digijava/kernel/validators/activity/ConditionalRequiredValidator.java]: Updated java doc and made the validator fail explicitly for translated fields.

[the remark this answers is not in the document]
Comment (nmandrescu) [reply amp/WEB-INF/src/org/digijava/kernel/validators/activity/ConditionalRequiredValidator.java]: Sure, we can leave to be worked together with required validator

File: amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java (hunk 2 above)
Comment (developster) [reply amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java]: If we're speaking of this particular test, then one violation is the correct answer.
If you mean to have a test which contains both kinds of violations at the same time then there is `#testInvalidLocationAndImplLoc`.

Comment (nmandrescu) [reply amp/src/test/java/org/digijava/kernel/validators/activity/ImplementationLevelValidatorTest.java]: You are right, I see it later in code

File: amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java [hunk 5]
```diff
@@ -627,9 +637,10 @@
                 discriminatorOption = CategoryConstants.PROJECT_CATEGORY_KEY, fmPath="/Activity Form/Identification/Project Category", pickIdOnly=true),
         @Interchangeable(fieldTitle = "Implementation Level", importable=true, multipleValues=false, 
                 discriminatorOption = CategoryConstants.IMPLEMENTATION_LEVEL_KEY, fmPath="/Activity Form/Location/Implementation Level", pickIdOnly=true),
-        @Interchangeable(fieldTitle = "Implementation Location", importable=true, multipleValues=false, 
-                discriminatorOption = CategoryConstants.IMPLEMENTATION_LOCATION_KEY, fmPath="/Activity Form/Location/Implementation Location", 
-                dependencies = {InterchangeDependencyResolver.IMPLEMENTATION_LOCATION_VALID_KEY}, pickIdOnly=true),
+        @Interchangeable(fieldTitle = "Implementation Location", importable = true, multipleValues = false,
```
Comment (nmandrescu) [inline amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: Please also use constants for implLoc and implLevel

Comment (nmandrescu) [reply amp/WEB-INF/src/org/digijava/module/aim/dbentity/AmpActivityFields.java]: Can be done in the next tasks

**Pull request merged**